File: cmd/drift/internal/templates/ios/StorageHandler.swift
/// StorageHandler.swift
/// Handles file system access and document picking for the Drift platform channel.

import UIKit
import UniformTypeIdentifiers

final class StorageHandler: NSObject {
    static let shared = StorageHandler()

    private var pendingSaveData: Data?
    private var pendingRequestType: String = "pickFile"
    private var pendingRequestId: String?

    private override init() {
        super.init()
    }

    static func handle(method: String, args: Any?) -> (Any?, Error?) {
        switch method {
        case "pickFile":
            return shared.pickFile(args: args)
        case "pickDirectory":
            return shared.pickDirectory(args: args)
        case "saveFile":
            return shared.saveFile(args: args)
        case "readFile":
            return readFile(args: args)
        case "writeFile":
            return writeFile(args: args)
        case "deleteFile":
            return deleteFile(args: args)
        case "getFileInfo":
            return getFileInfo(args: args)
        case "getAppDirectory":
            return getAppDirectory(args: args)
        default:
            return (nil, NSError(domain: "Storage", code: 404, userInfo: [NSLocalizedDescriptionKey: "Unknown method: \(method)"]))
        }
    }

    private func pickFile(args: Any?) -> (Any?, Error?) {
        let dict = args as? [String: Any] ?? [:]
        let allowMultiple = dict["allowMultiple"] as? Bool ?? false
        let allowedTypes = dict["allowedTypes"] as? [String] ?? ["public.item"]

        pendingRequestType = "pickFile"
        pendingRequestId = dict["requestId"] as? String

        DispatchQueue.main.async {
            self.presentDocumentPicker(allowMultiple: allowMultiple, allowedTypes: allowedTypes)
        }

        // Result will be delivered via drift/storage/result event channel
        return (["pending": true], nil)
    }

    private func presentDocumentPicker(allowMultiple: Bool, allowedTypes: [String]) {
        var contentTypes: [UTType] = []
        for typeString in allowedTypes {
            if let type = UTType(typeString) {
                contentTypes.append(type)
            } else if let type = UTType(mimeType: typeString) {
                contentTypes.append(type)
            }
        }

        if contentTypes.isEmpty {
            contentTypes = [.item]
        }

        let picker = UIDocumentPickerViewController(forOpeningContentTypes: contentTypes, asCopy: true)
        picker.allowsMultipleSelection = allowMultiple
        picker.delegate = self

        presentPicker(picker)
    }

    private func pickDirectory(args: Any?) -> (Any?, Error?) {
        let dict = args as? [String: Any] ?? [:]
        pendingRequestType = "pickDirectory"
        pendingRequestId = dict["requestId"] as? String

        DispatchQueue.main.async {
            let picker = UIDocumentPickerViewController(forOpeningContentTypes: [.folder])
            picker.delegate = self
            self.presentPicker(picker)
        }

        // Result will be delivered via drift/storage/result event channel
        return (["pending": true], nil)
    }

    private func saveFile(args: Any?) -> (Any?, Error?) {
        guard let dict = args as? [String: Any] else {
            return (nil, NSError(domain: "Storage", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid arguments"]))
        }

        let suggestedName = dict["suggestedName"] as? String ?? "file"

        guard let data = StorageHandler.extractData(from: dict["data"]) else {
            return (nil, NSError(domain: "Storage", code: 400, userInfo: [NSLocalizedDescriptionKey: "Missing data"]))
        }

        pendingRequestId = dict["requestId"] as? String

        // Save to temp file first, then present export picker
        let tempURL = FileManager.default.temporaryDirectory.appendingPathComponent(suggestedName)

        do {
            try data.write(to: tempURL)
            pendingSaveData = data

            DispatchQueue.main.async {
                let activityVC = UIActivityViewController(activityItems: [tempURL], applicationActivities: nil)
                activityVC.completionWithItemsHandler = { _, completed, _, error in
                    if completed {
                        self.sendSaveFileResult(tempURL.path)
                    } else if let error = error {
                        self.sendError("saveFile", message: error.localizedDescription)
                    } else {
                        self.sendCancelled("saveFile")
                    }
                    self.pendingSaveData = nil
                }

                if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
                   let rootVC = windowScene.windows.first?.rootViewController {
                    var topVC = rootVC
                    while let presented = topVC.presentedViewController {
                        topVC = presented
                    }

                    if let popover = activityVC.popoverPresentationController {
                        popover.sourceView = topVC.view
                        popover.sourceRect = CGRect(x: topVC.view.bounds.midX, y: topVC.view.bounds.midY, width: 0, height: 0)
                    }

                    topVC.present(activityVC, animated: true)
                }
            }

            // Result will be delivered via event channel
            return (["pending": true], nil)
        } catch {
            return (nil, error)
        }
    }

    private static func readFile(args: Any?) -> (Any?, Error?) {
        guard let dict = args as? [String: Any],
              let path = dict["path"] as? String else {
            return (nil, NSError(domain: "Storage", code: 400, userInfo: [NSLocalizedDescriptionKey: "Missing path"]))
        }

        do {
            let url = URL(fileURLWithPath: path)
            let data = try Data(contentsOf: url)
            return (["data": [UInt8](data)], nil)
        } catch {
            return (nil, error)
        }
    }

    private static func writeFile(args: Any?) -> (Any?, Error?) {
        guard let dict = args as? [String: Any],
              let path = dict["path"] as? String else {
            return (nil, NSError(domain: "Storage", code: 400, userInfo: [NSLocalizedDescriptionKey: "Missing path"]))
        }

        guard let data = extractData(from: dict["data"]) else {
            return (nil, NSError(domain: "Storage", code: 400, userInfo: [NSLocalizedDescriptionKey: "Missing data"]))
        }

        do {
            let url = URL(fileURLWithPath: path)
            try data.write(to: url)
            return (nil, nil)
        } catch {
            return (nil, error)
        }
    }

    private static func deleteFile(args: Any?) -> (Any?, Error?) {
        guard let dict = args as? [String: Any],
              let path = dict["path"] as? String else {
            return (nil, NSError(domain: "Storage", code: 400, userInfo: [NSLocalizedDescriptionKey: "Missing path"]))
        }

        do {
            let url = URL(fileURLWithPath: path)
            try FileManager.default.removeItem(at: url)
            return (nil, nil)
        } catch {
            return (nil, error)
        }
    }

    private static func getFileInfo(args: Any?) -> (Any?, Error?) {
        guard let dict = args as? [String: Any],
              let path = dict["path"] as? String else {
            return (nil, NSError(domain: "Storage", code: 400, userInfo: [NSLocalizedDescriptionKey: "Missing path"]))
        }

        let url = URL(fileURLWithPath: path)

        do {
            let attrs = try FileManager.default.attributesOfItem(atPath: path)
            let isDirectory = (attrs[.type] as? FileAttributeType) == .typeDirectory
            let size = attrs[.size] as? Int64 ?? 0
            let modificationDate = attrs[.modificationDate] as? Date

            var mimeType = ""
            if let uti = UTType(filenameExtension: url.pathExtension) {
                mimeType = uti.preferredMIMEType ?? ""
            }

            return ([
                "name": url.lastPathComponent,
                "path": path,
                "size": size,
                "mimeType": mimeType,
                "isDirectory": isDirectory,
                "lastModified": Int64((modificationDate?.timeIntervalSince1970 ?? 0) * 1000)
            ], nil)
        } catch {
            return (nil, error)
        }
    }

    private static func getAppDirectory(args: Any?) -> (Any?, Error?) {
        guard let dict = args as? [String: Any],
              let directory = dict["directory"] as? String else {
            return (nil, NSError(domain: "Storage", code: 400, userInfo: [NSLocalizedDescriptionKey: "Missing directory"]))
        }

        let fileManager = FileManager.default
        let path: String?

        switch directory {
        case "documents":
            path = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first?.path
        case "cache":
            path = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first?.path
        case "temp":
            path = NSTemporaryDirectory()
        case "support":
            path = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first?.path
        default:
            path = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first?.path
        }

        return (["path": path], nil)
    }

    private static func extractData(from value: Any?) -> Data? {
        if let data = value as? Data {
            return data
        }
        if let bytes = value as? [UInt8] {
            return Data(bytes)
        }
        if let bytes = value as? [Int] {
            return Data(bytes.map { UInt8(truncatingIfNeeded: $0) })
        }
        if let string = value as? String {
            return string.data(using: .utf8)
        }
        return nil
    }

    private func presentPicker(_ picker: UIViewController) {
        if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
           let rootVC = windowScene.windows.first?.rootViewController {
            var topVC = rootVC
            while let presented = topVC.presentedViewController {
                topVC = presented
            }
            topVC.present(picker, animated: true)
        }
    }

    // MARK: - Event Sending

    private func sendPickFileResult(_ files: [[String: Any]]) {
        var event: [String: Any] = [
            "type": "pickFile",
            "files": files
        ]
        if let reqId = pendingRequestId { event["requestId"] = reqId }
        PlatformChannelManager.shared.sendEvent(channel: "drift/storage/result", data: event)
        pendingRequestId = nil
    }

    private func sendPickDirectoryResult(_ path: String?) {
        var event: [String: Any] = [
            "type": "pickDirectory",
            "path": path as Any
        ]
        if let reqId = pendingRequestId { event["requestId"] = reqId }
        PlatformChannelManager.shared.sendEvent(channel: "drift/storage/result", data: event)
        pendingRequestId = nil
    }

    private func sendSaveFileResult(_ path: String) {
        var event: [String: Any] = [
            "type": "saveFile",
            "path": path
        ]
        if let reqId = pendingRequestId { event["requestId"] = reqId }
        PlatformChannelManager.shared.sendEvent(channel: "drift/storage/result", data: event)
        pendingRequestId = nil
    }

    private func sendCancelled(_ requestType: String) {
        var event: [String: Any] = [
            "type": requestType,
            "cancelled": true
        ]
        if let reqId = pendingRequestId { event["requestId"] = reqId }
        PlatformChannelManager.shared.sendEvent(channel: "drift/storage/result", data: event)
        pendingRequestId = nil
    }

    private func sendError(_ requestType: String, message: String) {
        var event: [String: Any] = [
            "type": requestType,
            "error": message
        ]
        if let reqId = pendingRequestId { event["requestId"] = reqId }
        PlatformChannelManager.shared.sendEvent(channel: "drift/storage/result", data: event)
        pendingRequestId = nil
    }
}

// MARK: - UIDocumentPickerDelegate

extension StorageHandler: UIDocumentPickerDelegate {
    func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        var files: [[String: Any]] = []

        for url in urls {
            // Start accessing security-scoped resource
            let accessing = url.startAccessingSecurityScopedResource()
            defer {
                if accessing {
                    url.stopAccessingSecurityScopedResource()
                }
            }

            do {
                let attrs = try FileManager.default.attributesOfItem(atPath: url.path)
                let isDirectory = (attrs[.type] as? FileAttributeType) == .typeDirectory
                let size = attrs[.size] as? Int64 ?? 0

                var mimeType = ""
                if let uti = UTType(filenameExtension: url.pathExtension) {
                    mimeType = uti.preferredMIMEType ?? ""
                }

                files.append([
                    "name": url.lastPathComponent,
                    "path": url.path,
                    "uri": url.absoluteString,
                    "mimeType": mimeType,
                    "size": size,
                    "isDirectory": isDirectory
                ])
            } catch {
                // Skip files we can't read
            }
        }

        // Check if this was a directory picker (single folder selection)
        if files.count == 1, let file = files.first, file["isDirectory"] as? Bool == true {
            sendPickDirectoryResult(file["path"] as? String)
        } else {
            sendPickFileResult(files)
        }
    }

    func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
        sendCancelled(pendingRequestType)
    }
}

File: cmd/drift/internal/templates/android/java/StorageHandler.kt
/**
 * StorageHandler.kt
 * Handles file system access and document picking for the Drift platform channel.
 */
package {{.PackageName}}

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Environment
import android.provider.DocumentsContract
import android.provider.OpenableColumns
import android.webkit.MimeTypeMap
import java.io.File
import android.util.Base64

object StorageHandler {
    private const val PICK_FILE_REQUEST = 9002
    private const val PICK_DIRECTORY_REQUEST = 9003
    private const val SAVE_FILE_REQUEST = 9004

    private var pendingSaveData: ByteArray? = null
    private var pendingRequestType: String? = null
    private var pendingRequestId: String? = null

    fun handle(context: Context, method: String, args: Any?): Pair<Any?, Exception?> {
        return when (method) {
            "pickFile" -> pickFile(context, args)
            "pickDirectory" -> pickDirectory(context, args)
            "saveFile" -> saveFile(context, args)
            "readFile" -> readFile(context, args)
            "writeFile" -> writeFile(context, args)
            "deleteFile" -> deleteFile(context, args)
            "getFileInfo" -> getFileInfo(context, args)
            "getAppDirectory" -> getAppDirectory(context, args)
            else -> Pair(null, IllegalArgumentException("Unknown method: $method"))
        }
    }

    private fun pickFile(context: Context, args: Any?): Pair<Any?, Exception?> {
        val activity = PlatformChannelManager.currentActivity()
            ?: return Pair(null, IllegalStateException("No activity available"))

        val argsMap = args as? Map<*, *> ?: emptyMap<String, Any>()
        val allowMultiple = argsMap["allowMultiple"] as? Boolean ?: false
        @Suppress("UNCHECKED_CAST")
        val allowedTypes = argsMap["allowedTypes"] as? List<String> ?: listOf("*/*")

        pendingRequestType = "pickFile"
        pendingRequestId = argsMap["requestId"] as? String

        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
            addCategory(Intent.CATEGORY_OPENABLE)
            type = if (allowedTypes.size == 1) allowedTypes[0] else "*/*"
            if (allowedTypes.size > 1) {
                putExtra(Intent.EXTRA_MIME_TYPES, allowedTypes.toTypedArray())
            }
            if (allowMultiple) {
                putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)
            }
        }

        activity.startActivityForResult(intent, PICK_FILE_REQUEST)
        // Result will be delivered via drift/storage/result event channel
        return Pair(mapOf("pending" to true), null)
    }

    private fun pickDirectory(context: Context, args: Any?): Pair<Any?, Exception?> {
        val activity = PlatformChannelManager.currentActivity()
            ?: return Pair(null, IllegalStateException("No activity available"))

        val argsMap = args as? Map<*, *> ?: emptyMap<String, Any>()
        pendingRequestType = "pickDirectory"
        pendingRequestId = argsMap["requestId"] as? String

        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT_TREE)
        activity.startActivityForResult(intent, PICK_DIRECTORY_REQUEST)
        // Result will be delivered via drift/storage/result event channel
        return Pair(mapOf("pending" to true), null)
    }

    private fun saveFile(context: Context, args: Any?): Pair<Any?, Exception?> {
        val activity = PlatformChannelManager.currentActivity()
            ?: return Pair(null, IllegalStateException("No activity available"))

        val argsMap = args as? Map<*, *>
            ?: return Pair(null, IllegalArgumentException("Invalid arguments"))

        val suggestedName = argsMap["suggestedName"] as? String ?: "file"
        val mimeType = argsMap["mimeType"] as? String ?: "*/*"

        @Suppress("UNCHECKED_CAST")
        val data = when (val dataArg = argsMap["data"]) {
            is ByteArray -> dataArg
            is String -> Base64.decode(dataArg, Base64.DEFAULT)
            is List<*> -> (dataArg as List<Number>).map { it.toByte() }.toByteArray()
            else -> null
        }

        if (data == null) {
            return Pair(null, IllegalArgumentException("Missing data"))
        }

        pendingSaveData = data
        pendingRequestType = "saveFile"
        pendingRequestId = argsMap["requestId"] as? String

        val intent = Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
            addCategory(Intent.CATEGORY_OPENABLE)
            type = mimeType
            putExtra(Intent.EXTRA_TITLE, suggestedName)
        }

        activity.startActivityForResult(intent, SAVE_FILE_REQUEST)
        // Result will be delivered via drift/storage/result event channel
        return Pair(mapOf("pending" to true), null)
    }

    private fun readFile(context: Context, args: Any?): Pair<Any?, Exception?> {
        val argsMap = args as? Map<*, *>
            ?: return Pair(null, IllegalArgumentException("Invalid arguments"))
        val path = argsMap["path"] as? String
            ?: return Pair(null, IllegalArgumentException("Missing path"))

        return try {
            val data = if (path.startsWith("content://")) {
                val uri = Uri.parse(path)
                context.contentResolver.openInputStream(uri)?.use { it.readBytes() }
            } else {
                File(path).readBytes()
            }
            Pair(mapOf("data" to data), null)
        } catch (e: Exception) {
            Pair(null, e)
        }
    }

    private fun writeFile(context: Context, args: Any?): Pair<Any?, Exception?> {
        val argsMap = args as? Map<*, *>
            ?: return Pair(null, IllegalArgumentException("Invalid arguments"))
        val path = argsMap["path"] as? String
            ?: return Pair(null, IllegalArgumentException("Missing path"))

        @Suppress("UNCHECKED_CAST")
        val data = when (val dataArg = argsMap["data"]) {
            is ByteArray -> dataArg
            is String -> Base64.decode(dataArg, Base64.DEFAULT)
            is List<*> -> (dataArg as List<Number>).map { it.toByte() }.toByteArray()
            else -> null
        } ?: return Pair(null, IllegalArgumentException("Missing data"))

        return try {
            if (path.startsWith("content://")) {
                val uri = Uri.parse(path)
                context.contentResolver.openOutputStream(uri)?.use { it.write(data) }
            } else {
                File(path).writeBytes(data)
            }
            Pair(null, null)
        } catch (e: Exception) {
            Pair(null, e)
        }
    }

    private fun deleteFile(context: Context, args: Any?): Pair<Any?, Exception?> {
        val argsMap = args as? Map<*, *>
            ?: return Pair(null, IllegalArgumentException("Invalid arguments"))
        val path = argsMap["path"] as? String
            ?: return Pair(null, IllegalArgumentException("Missing path"))

        return try {
            if (path.startsWith("content://")) {
                val uri = Uri.parse(path)
                DocumentsContract.deleteDocument(context.contentResolver, uri)
            } else {
                File(path).delete()
            }
            Pair(null, null)
        } catch (e: Exception) {
            Pair(null, e)
        }
    }

    private fun getFileInfo(context: Context, args: Any?): Pair<Any?, Exception?> {
        val argsMap = args as? Map<*, *>
            ?: return Pair(null, IllegalArgumentException("Invalid arguments"))
        val path = argsMap["path"] as? String
            ?: return Pair(null, IllegalArgumentException("Missing path"))

        return try {
            val info = if (path.startsWith("content://")) {
                getContentUriInfo(context, Uri.parse(path))
            } else {
                getFilePathInfo(path)
            }
            Pair(info, null)
        } catch (e: Exception) {
            Pair(null, e)
        }
    }

    private fun getAppDirectory(context: Context, args: Any?): Pair<Any?, Exception?> {
        val argsMap = args as? Map<*, *>
            ?: return Pair(null, IllegalArgumentException("Invalid arguments"))
        val directory = argsMap["directory"] as? String
            ?: return Pair(null, IllegalArgumentException("Missing directory"))

        val path = when (directory) {
            "documents" -> context.getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS)?.absolutePath
            "cache" -> context.cacheDir.absolutePath
            "temp" -> context.cacheDir.absolutePath + "/temp"
            "support" -> context.filesDir.absolutePath
            else -> context.filesDir.absolutePath
        }

        return Pair(mapOf("path" to path), null)
    }

    private fun getContentUriInfo(context: Context, uri: Uri): Map<String, Any?> {
        var name = ""
        var size = 0L
        var mimeType = context.contentResolver.getType(uri) ?: ""

        context.contentResolver.query(uri, null, null, null, null)?.use { cursor ->
            if (cursor.moveToFirst()) {
                val nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                val sizeIndex = cursor.getColumnIndex(OpenableColumns.SIZE)
                if (nameIndex >= 0) name = cursor.getString(nameIndex) ?: ""
                if (sizeIndex >= 0 && !cursor.isNull(sizeIndex)) size = cursor.getLong(sizeIndex)
            }
        }

        return mapOf(
            "name" to name,
            "path" to uri.toString(),
            "uri" to uri.toString(),
            "size" to size,
            "mimeType" to mimeType,
            "isDirectory" to false,
            "lastModified" to 0L
        )
    }

    private fun getFilePathInfo(path: String): Map<String, Any?> {
        val file = File(path)
        val mimeType = MimeTypeMap.getSingleton()
            .getMimeTypeFromExtension(file.extension) ?: ""

        return mapOf(
            "name" to file.name,
            "path" to file.absolutePath,
            "size" to file.length(),
            "mimeType" to mimeType,
            "isDirectory" to file.isDirectory,
            "lastModified" to file.lastModified()
        )
    }

    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?, context: Context) {
        when (requestCode) {
            PICK_FILE_REQUEST -> {
                if (resultCode == Activity.RESULT_OK) {
                    val files = mutableListOf<Map<String, Any?>>()
                    data?.clipData?.let { clipData ->
                        for (i in 0 until clipData.itemCount) {
                            clipData.getItemAt(i).uri?.let { uri ->
                                files.add(getContentUriInfo(context, uri))
                            }
                        }
                    } ?: data?.data?.let { uri ->
                        files.add(getContentUriInfo(context, uri))
                    }
                    sendPickFileResult(files)
                } else {
                    sendCancelled("pickFile")
                }
                pendingRequestType = null
            }
            PICK_DIRECTORY_REQUEST -> {
                if (resultCode == Activity.RESULT_OK) {
                    val path = data?.data?.toString()
                    sendPickDirectoryResult(path)
                } else {
                    sendCancelled("pickDirectory")
                }
                pendingRequestType = null
            }
            SAVE_FILE_REQUEST -> {
                if (resultCode == Activity.RESULT_OK) {
                    data?.data?.let { uri ->
                        pendingSaveData?.let { saveData ->
                            try {
                                context.contentResolver.openOutputStream(uri)?.use {
                                    it.write(saveData)
                                }
                                sendSaveFileResult(uri.toString())
                            } catch (e: Exception) {
                                sendError("saveFile", e.message ?: "Failed to save file")
                            }
                        } ?: sendError("saveFile", "No data to save")
                    } ?: sendCancelled("saveFile")
                } else {
                    sendCancelled("saveFile")
                }
                pendingSaveData = null
                pendingRequestType = null
            }
        }
    }

    private fun sendPickFileResult(files: List<Map<String, Any?>>) {
        val event = mutableMapOf<String, Any?>(
            "type" to "pickFile",
            "files" to files
        )
        pendingRequestId?.let { event["requestId"] = it }
        PlatformChannelManager.sendEvent("drift/storage/result", event)
        pendingRequestId = null
    }

    private fun sendPickDirectoryResult(path: String?) {
        val event = mutableMapOf<String, Any?>(
            "type" to "pickDirectory",
            "path" to path
        )
        pendingRequestId?.let { event["requestId"] = it }
        PlatformChannelManager.sendEvent("drift/storage/result", event)
        pendingRequestId = null
    }

    private fun sendSaveFileResult(path: String) {
        val event = mutableMapOf<String, Any?>(
            "type" to "saveFile",
            "path" to path
        )
        pendingRequestId?.let { event["requestId"] = it }
        PlatformChannelManager.sendEvent("drift/storage/result", event)
        pendingRequestId = null
    }

    private fun sendCancelled(requestType: String) {
        val event = mutableMapOf<String, Any?>(
            "type" to requestType,
            "cancelled" to true
        )
        pendingRequestId?.let { event["requestId"] = it }
        PlatformChannelManager.sendEvent("drift/storage/result", event)
        pendingRequestId = null
    }

    private fun sendError(requestType: String, message: String) {
        val event = mutableMapOf<String, Any?>(
            "type" to requestType,
            "error" to message
        )
        pendingRequestId?.let { event["requestId"] = it }
        PlatformChannelManager.sendEvent("drift/storage/result", event)
        pendingRequestId = null
    }
}
